tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1497 @@
+    "/tensorstore/kvstore/s3/bytes_written",
+    "Bytes written by the s3 kvstore driver");
+
+auto& s3_retries = internal_metrics::Counter<int64_t>::New(
+    "/tensorstore/kvstore/s3/retries",
+    "Count of all retried S3 requests (read/write/delete)");
+
+auto& s3_read = internal_metrics::Counter<int64_t>::New(
+    "/tensorstore/kvstore/s3/read", "S3 driver kvstore::Read calls");
+
+auto& s3_read_latency_ms =
+    internal_metrics::Histogram<internal_metrics::DefaultBucketer>::New(
+        "/tensorstore/kvstore/s3/read_latency_ms",
+        "S3 driver kvstore::Read latency (ms)");
+
+auto& s3_write = internal_metrics::Counter<int64_t>::New(
+    "/tensorstore/kvstore/s3/write", "S3 driver kvstore::Write calls");
+
+auto& s3_write_latency_ms =
+    internal_metrics::Histogram<internal_metrics::DefaultBucketer>::New(
+        "/tensorstore/kvstore/s3/write_latency_ms",
+        "S3 driver kvstore::Write latency (ms)");
+
+auto& s3_delete_range = internal_metrics::Counter<int64_t>::New(
+    "/tensorstore/kvstore/s3/delete_range",
+    "S3 driver kvstore::DeleteRange calls");
+
+auto& s3_list = internal_metrics::Counter<int64_t>::New(
+    "/tensorstore/kvstore/s3/list", "S3 driver kvstore::List calls");
+
+static constexpr char kEmptySha256[] =
+  "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
+
+
+/// Adds the generation header to the provided builder.
+bool AddGenerationHeader(S3RequestBuilder * builder,
+                         std::string_view header,
+                         const StorageGeneration& gen) {
+  if (StorageGeneration::IsUnknown(gen)) {
+    // Unconditional.
+    return false;
+  } else {
+    // One of two cases applies:
+    //
+    // 1. `gen` is a `StorageGeneration::FromString("etag;last_modified")` generation.
+    //    The condition is specified as `=etag`.
+
+    auto [etag, last_mod] = ExtractETagAndLastModified(gen);
+    builder->AddHeader(absl::StrCat(header, ": ", etag));
+    return true;
+  }
+}
+
+
+std::pair<std::string, std::string> payload_sha256(const absl::Cord & cord=absl::Cord()) {
+  SHA256Digester sha256;
+  sha256.Write(cord);
+  auto digest = sha256.Digest();
+  auto digest_sv = std::string_view(
+    reinterpret_cast<const char *>(&digest[0]),
+    digest.size());
+
+  std::string base64_result;
+  absl::Base64Escape(digest_sv, &base64_result);
+
+  return {absl::BytesToHexString(digest_sv), base64_result};
+}
+
+std::pair<std::string, std::string> payload_md5(const absl::Cord & cord=absl::Cord()) {
+  MD5Digester md5;
+  md5.Write(cord);
+  auto digest = md5.Digest();
+  auto digest_sv = std::string_view(
+    reinterpret_cast<const char *>(&digest[0]),
+    digest.size());
+
+  std::string base64_result;
+  absl::Base64Escape(digest_sv, &base64_result);
+  return {base64_result, absl::BytesToHexString(digest_sv)};
+
+}
+
+struct S3KeyValueStoreSpecData {
+  std::string bucket;
+  bool requester_pays;
+  std::string endpoint;
```

Comment:
This should be std::optional<std::string> instead

Reply:
changed to `std::optional<std::string>`